### PR TITLE
Fix nested class mutability and json serialization

### LINKED ISSUE #45: to_json_serializable_dict doesn't correctly serialize nested classes
This happens when enclosing classes set the nested class param value within its constructor

Failure mode:
```python
class A(Params):
    f = FloatParam(help='float', default=2)

class B(Params):
    a = A(f = 5)

b = B()
to_json_serializable_dict(b)  # will give {'a': {'f': 5, ...}, ...} as expected
b.a.f = 7
to_json_serializable_dict(b)  # will give {'a': {'f': 5, ...}, ...}, which should actually be
# {'a': {'f': 7, ...}, ...}
```

### LINKED ISSUE #46: Nested params classes are mutable and shouldn't be
Ex.
```python

class A(Params):
    f = FloatParam(help='float', default=2)

class B(Params):
    a = A(f = 5)

b = B()
b.a.f = 7
x = B()
print(x.a.f)  # prints 7 and not 5

### PR DESCRIPTION
**Pull request by schuylerfried**
closes #45 
closes #46

### CODE REVIEW

File: paranormal/parameter_interface.py [hunk 1]
```diff
@@ -86,6 +88,15 @@ def si_set(self, param_name: str, value: Union[float, int, Iterable]):
             unit = get_param_unit(type(self), param_name)
             setattr(self, param_name, unconvert_si_units(value, unit))
 
+    def si_update(self, **kwargs) -> None:
```
Comment (deanna-abrams) [inline paranormal/parameter_interface.py]: don't love having two ways to si set depending on how you want to use it. can we consolidate to one or the other and have it handle both cases?

File: paranormal/parameter_interface.py [hunk 2]
```diff
@@ -119,6 +130,15 @@ def _ensure_properties_are_working(params: Params) -> None:
             getattr(params, k)
 
 
+def _copy_nested_classes(params: Params):
+    """
+    Deepcopy and set any nested params classes to make sure users cannot mutate the nested classes
```
Comment (deanna-abrams) [inline paranormal/parameter_interface.py]: specify mutate defaults?

File: paranormal/test/test_parameter_interface.py [hunk 3]
```diff
@@ -157,6 +157,25 @@ def test_json_serialization():
     j = json.loads(json.dumps(to_json_serializable_dict(y)))
     assert y == from_json_serializable_dict(j)
 
+    # test with setting nested params
```
Comment (deanna-abrams) [inline paranormal/test/test_parameter_interface.py]: more specific: `# test changing defaults of nested params in one class instance doesn't change the default for all class instances`

File: paranormal/parameter_interface.py (hunk 1 above)
Comment (schuylerfried) [reply paranormal/parameter_interface.py]: I'd actually argue that this is very similar to the python dictionary: `x['5'] = 45` and `x.update('5'=45, '10'=...)`, which is a little redundant, but doesn't seem to hurt. In this case, si_update is calling si_set, which isn't really duplicating any code either..

Comment (deanna-abrams) [reply paranormal/parameter_interface.py]: hmm. these are good points - works for me

**Pull request merged**